### DISCUSSION TITLE
How does one run this?

### DISCUSSION THREAD

**Discussion by GoWithitRoger** [n0rt0nthec4t/Nest_accfactory#7] [Q&A] [answered]
Apologies for the very basic question, but other than Docker I'm not seeing how/where this is to be run. 

If Docker is the answer, I'm having an issue setting that up. I think it's around the persistent volumes for data and configuration — I run the following (on Mac, with Docker desktop installed, in terminal): 

```
docker run --name nest_accfactory\
    --restart always\
    --volume /Users/myusername/GoogCamDocker/conf:/opt/Nest_accfactory/conf:ro\
    --volume /Users/myusername/GoogCamDocker/persist:/opt/Nest_accfactory/persist:rw\
    --network host\
    --detach n0rt0nthec4t/nest_accfactory
```

Then, I see that it starts in the Docker Dashboard...and then fails (and continuously restarts), with the logged error:

```
Specified configuration '/opt/Nest_accfactory/conf/Nest_config.json' cannot be found
Exiting.
```

**Comment by n0rt0nthec4t** [comment]
First up, I've not tested running with docker desktop on Mac, even though I develop on a Mac, so your milage might vary. Have your created the configuration file in your persistent volume folder on the Mac? It would be called "Nest_config.json" and be located in "/Users/myusername/GoogCamDocker/conf"

**Reply by GoWithitRoger** [reply]
Well I really appreciate the very quick reply.

I didn't pick up from the docker instructions that _I_ was to create the .json file; I thought we were simply supposed to make the directory available. I just stuck a blank "Nest_config.json" in that location and I see in the log:
```
Configuration file '/opt/Nest_accfactory/conf/Nest_config.json' contains invalid options
Exiting.
```

So that's progress. I take it I should paste in something like this (from the Readme)?
```
{
    "GoogleToken" : {
        "issuetoken" : "<google issue token url>",
        "cookie" : "<google cookie>"
    },
    "HKSV" : true,
    "H264Encoder" : "copy"
}
```

**Comment by n0rt0nthec4t** [comment]
Yep.

**Reply by GoWithitRoger** [reply]
Alright! Still not working, but I've got it attempting to authenticate in with my Nest session token. 
```
2023-11-27 09:18:06 2023-11-27 15:18:06 [Nest] Starting /opt/Nest_accfactory/Nest_accfactory.js using HAP-NodeJS library v0.11.1
2023-11-27 09:18:06 2023-11-27 15:18:06 [Nest] Configuration will be read from '/opt/Nest_accfactory/conf/Nest_config.json'
2023-11-27 09:18:06 2023-11-27 15:18:06 [Nest] Performing Nest account authorisation
2023-11-27 09:18:07 2023-11-27 15:18:07 [Nest] Account authorisation failed
```

Using this config (I've edited out most of the very-long token):
```
{
  "SessionToken": "g.0.ey...5A",
  "HKSV": true,
  "H264Encoder": "copy"
}
```

I'm fairly confident the issue is that my account is a Field Test (FT) account. My Nest web app is at home.ft.nest.com (and for what it's worth my access token doesn't start with a "b"). 

The [home bridge-nest](https://github.com/chrisjshull/homebridge-nest/tree/999faba0407c4560c31f6c1e04756138c596daa9) Homebriedge add-on has a setting to enable use with FT account, but I can't tell from the code what it does differently to navigate that. Any idea how to get yours working with a FT account?



**First Edit/update**: I realized that I migrated my account some time ago, so I don't use the "Sign in with Nest" option at home.ft.nest.com, but instead use the Sign in with Google method. I grabbed the issue token and cookie for the Google sign in method. And I get the following error:

```
2023-11-27 09:41:45 2023-11-27 15:41:45 [Nest] Starting /opt/Nest_accfactory/Nest_accfactory.js using HAP-NodeJS library v0.11.1
2023-11-27 09:41:45 2023-11-27 15:41:45 [Nest] Configuration will be read from '/opt/Nest_accfactory/conf/Nest_config.json'
2023-11-27 09:41:45 2023-11-27 15:41:45 [Nest] Performing Google account authorisation
```

That's using this config (I've edited out most of the token and cookie):
```
{
    "GoogleToken" : {
        "issuetoken" : "https://accounts.google.com/o/oauth2/iframerpc?action=issueToken&response_type=token%20id_token&login_hint=AJ...https%3A%2F%2Fhome.ft.nest.com&scope=openid%20profile%20email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fnest-account&ss_domain=https%3A%2F%2Fhome.ft.nest.com&include_granted_scopes=true&auto=0",
        "cookie" : "LSOLH=_SVI...pJ"
    },
    "HKSV" : true,
    "H264Encoder" : "copy"
}
```

### Second, bigger edit/update

1. I found (at least some of the) accommodations made in that "homebridge-nest" addon that look like they allow compatibility with field test cameras. 
2. And then --keeping in mind I don't know what I'm doing-- I forked your repo here and made [Nest_accfactory_ft](https://github.com/GoWithitRoger/Nest_accfactory_ft) where I've changed all of the `home.nest.com` instances I could find to `home.ft.nest.com'. 
3. Then, with a lot of google searching I was able to get docker to build an image from that repo and push the image to docker hub - https://hub.docker.com/repositories/rugdugit. There are actually two builds there, one ":latest" from my M1 MacBook, and another ":latestintel" after realizing I want this to run on an older Mac I use as my HomeBridge machine.

And now some new errors :)

```
2023-11-27 12:43:24 2023-11-27 18:43:24 [Nest] Starting /opt/Nest_accfactory/Nest_accfactory.js using HAP-NodeJS library v0.11.1
2023-11-27 12:43:24 2023-11-27 18:43:24 [Nest] The FFmpeg binary '/opt/Nest_accfactory/node_modules/ffmpeg-for-homebridge/ffmpeg' does not support the required libraries for doorbell and/or camera usage
2023-11-27 12:43:24 2023-11-27 18:43:24 [Nest] FFmpeg needs to be complied to include the following libraries: [ 'libspeex', 'libx264', 'libfdk-aac' ]
2023-11-27 12:43:24 2023-11-27 18:43:24 [Nest] Exiting.
```

I have a vague understanding of what this means, but no idea whether the local machine simply needs those dependencies installed, or if they need to be specified in the code for the docker image build, or...

Am I going about this the wrong way - is Docker the easiest way to run your accessory? Can I just run it in HomeBridge? (I think HAP-NodeJS is what HomeBridge is built on? But I don't see a way to run HAP-NodeJS on its own on my Mac, now how to then use your accessory with it.)

**Comment by n0rt0nthec4t** [comment]
It could be quite easy to add a configuration option for FT support, as have no one until now using an FT account. The FFmpeg binary issue is interesting. I might spin up a docker container on my Mac today to see what happens. I have an "alpha" version of the code in testing, but its x86/64 only atm

**Reply by GoWithitRoger** [reply]
If there's an easier way to run this than Docker, I'm all for it. I've just struck out on figuring out how else it can be run.

**Comment by n0rt0nthec4t** [comment, marked as answer]
Lets take this offline to email if like. Drop me a aline and we'll try some test code
